File: leetcode/1001-1999/1071-greatest-common-divisor-of-strings/GreatestCommonDivisorOfStrings.ts
function gcdOfStrings(s1: string, s2: string): string {
  if (s1 + s2 != s2 + s1) {
    return "";
  }

  const gcd = (a: number, b: number): number => {
    while (b !== 0) {
      [a, b] = [b, a % b];
    }

    return a;
  }

  return s1.substring(0, gcd(s1.length, s2.length));
};

console.log(gcdOfStrings("ABCABC", "ABC"));

File: leetcode/1001-2000/1001-1100/1071-1080/1071-greatest-common-divisor-of-strings/GreatestCommonDivisorOfStrings.java
public class GreatestCommonDivisorOfStrings {
  public static void main(String[] args) {
    System.out.println((new Solution()).gcdOfStrings("ABCABC", "ABC"));
  }
}

class Solution {
  public int gcd(int a, int b) {
    if (b == 0) {
      return a;
    }

    return gcd(b, a % b);
  }

  public String gcdOfStrings(String str1, String str2) {
    if (!(str1 + str2).equals(str2 + str1)) {
      return "";
    }

    return str1.substring(0, gcd(str1.length(), str2.length()));
  }
}
